sky/spot/recovery_strategy.py

```diff
@@ -255,18 +255,19 @@ def __init__(self, cluster_name: str, backend: 'backends.Backend',
                  task: 'task_lib.Task', retry_until_up: bool) -> None:
         super().__init__(cluster_name, backend, task, retry_until_up)
         # Note down the cloud/region of the launched cluster, so that we can
-        # first retry in the same cloud/region. (we cannot use handle, as it
-        # can be None if the cluster is preempted)
-        self.launched_cloud_region = None
+        # first retry in the same cloud/region. (Inside recover() we may not
+        # rely on cluster handle, as it can be None if the cluster is
+        # preempted.)
+        self._launched_cloud_region = None
 
-    def launch(self) -> Optional[float]:
-        launch_time = super().launch()
+    def _launch(self, max_retry=3, raise_on_failure=True) -> Optional[float]:
+        launch_time = super()._launch(max_retry, raise_on_failure)
```

Comment:
1. Document in _launch() a `Returns:` block? In particular it may return None if it failed to launch.
2. Can we guard against None here?

Reply:
Oops, spoke too soon. `raise_on_failure=True` means it's return-or-die. Still a Returns: block will be helpful!

Reply:
Good catch! Add the guard for the launch, otherwise, the assert will fail when the launch fails even if raise_on_failure is `False`.